In [197]:
import gensim.downloader as api

# Load pre-trained GloVe model
model = api.load("glove-wiki-gigaword-100")

In [213]:
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
import pandas as pd


#df = pd.read_csv("C:/Users/chiar/Downloads/final500 (2).csv")
df = pd.read_csv("C:/Users/chiar/Downloads/final10000.csv")
#df = df.dropna(subset=['Description'])
df =  df.dropna(subset=['review/text'])            
df = df.dropna(subset=['num_pages'])
df = df.drop(columns=['nan'])
df = df.drop(columns=['Unnamed: 0'])
df = df.drop(df[df["review/text"] == "['nan']"].index)
df = df.drop(columns=['Description'])
#df = df.drop(columns=['review/text'])
#df.to_csv("final_1365books.csv")

df

Title  \
3                        And Then There Were None   
6            The Lion, the Witch and the Wardrobe   
10        Harry Potter and the Chamber of Secrets   
17                         The Catcher in the Rye   
22                                         Lolita   
...                                           ...   
9431                               The Human Mind   
9433                            This Little Chick   
9452    View with a Grain of Sand: Selected Poems   
9485                                   Rogue Male   
9487  Stokes Field Guide to Birds: Western Region   

                                                 Author  \
3                                       Agatha Christie   
6                                           C. S. Lewis   
10                                        J. K. Rowling   
17                                       J. D. Salinger   
22                                     Vladimir Nabokov   
...                                                 ...   
9431                                     Robert Winston   
9433                                      John Lawrence   
9452  Wisława Szymborska,Stanisław Barańczak,Clar...   
9485                                 Geoffrey Household   
9487                       Donald Stokes,Lillian Stokes   

                                                  Genre  num_pages  \
3                                               Mystery      264.0   
6                           Fantasy, Children's fiction      206.0   
10                                              Fantasy      352.0   
17                                        Coming-of-age      234.0   
22                                                Novel      335.0   
...                                                 ...        ...   
9431  Nonfiction,Psychology,Science,Biology,Neurosci...      528.0   
9433  Childrens,Picture Books,Animals,Childrens,Stor...       32.0   
9452  Poetry,Cultural,Poland,European Literature,Pol...      214.0   
9485  Fiction,Thriller,Mystery,Crime,Mystery,Adventu...      182.0   
9487  Animals,Birds,Environment,Nature,Nonfiction,Re...      560.0   

                                            review/text  
3     ['book', 'one', 'ten', 'read', 'mystery', 'kil...  
6     ['lewis', 'world', 'witch', 'fantasy', 'wardro...  
10    ['book', 'potter', 'harry', 'fiction', 'scienc...  
17    ['amp', 'book', 'reading', 'read', 'biography'...  
22    ['lolita', 'humbert', 'language', 'nabokov', '...  
...                                                 ...  
9431  ['brain', 'winston', 'book', 'information', 's...  
9433  ['little', 'quot', 'chick', 'animal', 'reading...  
9452  ['poems', 'read', 'poetry', 'really', 'book', ...  
9485  ['book', 'country', 'british', 'tale', 'like',...  
9487  ['book', 'take', 'ca', 'birds', 'digital', 'ab...  

[1365 rows x 5 columns]

In [214]:

img = pd.read_csv("C:/Users/chiar/Downloads/moods (1).csv")
img["image"] = [0, 1, 2, 3, 4, 5, 6, 7, 8]

#getting the summary vector for the books' descriptions
for index, row in df.iterrows():
    review_text = row['review/text']
    if review_text != str(['nan']):
        review_text = review_text.strip("[] ' ' \"  ")
        review_text = review_text.replace("'", "")
        review_text = review_text.replace(" ", "")
        li = list(review_text.split(","))
        li = [word for word in li if word in model]
        summary_vector = sum(model[word] for word in li) / len(li)
        df.at[index, 'review/text'] = summary_vector
df 

#getting the summary vector for the images descriptions
for index, row in img.iterrows():
    review_text = row['vector']
    if review_text != str(['nan']):
        li = list(review_text.split(","))
        li = [word for word in li if word in model]
        summary_vector = sum(model[word] for word in li) / len(li)
        img.at[index, 'vector'] = summary_vector
img

def create_similarity_matrix(book_description_vectors, mood_vectors):
    similarity_matrix = np.zeros((len(book_description_vectors), len(mood_vectors)))
    for i, book_vector in enumerate(book_description_vectors):
        for j, mood_vector in enumerate(mood_vectors):
            if isinstance(book_vector, str) and book_vector == "['nan']":
                similarity_matrix[i, j] = 0
            else:
                #this check now becomes useless because we removed those earlier
                if np.isnan(book_vector).any() or np.isnan(mood_vector).any():
                    similarity_matrix[i, j] = 0
                else:
                    similarity_score = cosine_similarity([book_vector], [mood_vector])[0][0]
                    similarity_matrix[i, j] = similarity_score
    return similarity_matrix

mood_vectors = img['vector']
book_vectors = df['review/text']
book_titles = df['Title'].tolist()
similarity_matrix = create_similarity_matrix(book_vectors, mood_vectors)

matrix = pd.DataFrame(similarity_matrix, index=book_titles, columns=["happy", "sad", "misterious", "spring", "adventurous", "thought-provoking", "romance", "cozy", "fun"])
matrix

non_zero_rows = matrix[(matrix > 0.0000000).any(axis=1)]
#non_zero_rows.to_csv("mood_matrix.csv")

non_zero_rows

happy       sad  misterious  \
And Then There Were None                     0.708539  0.460718    0.367791   
The Lion, the Witch and the Wardrobe         0.593243  0.403501    0.364485   
Harry Potter and the Chamber of Secrets      0.661139  0.474719    0.517526   
The Catcher in the Rye                       0.679209  0.488919    0.411122   
Lolita                                       0.651534  0.456875    0.409290   
...                                               ...       ...         ...   
The Human Mind                               0.582039  0.359734    0.391833   
This Little Chick                            0.763816  0.511183    0.381383   
View with a Grain of Sand: Selected Poems    0.697610  0.460941    0.401551   
Rogue Male                                   0.627084  0.436670    0.476129   
Stokes Field Guide to Birds: Western Region  0.705235  0.454139    0.391497   

                                               spring  adventurous  \
And Then There Were None                     0.415213     0.475417   
The Lion, the Witch and the Wardrobe         0.361698     0.603045   
Harry Potter and the Chamber of Secrets      0.359339     0.617930   
The Catcher in the Rye                       0.377003     0.483698   
Lolita                                       0.351022     0.495811   
...                                               ...          ...   
The Human Mind                               0.341313     0.435736   
This Little Chick                            0.454744     0.469397   
View with a Grain of Sand: Selected Poems    0.372800     0.474141   
Rogue Male                                   0.332246     0.555670   
Stokes Field Guide to Birds: Western Region  0.342610     0.538794   

                                             thought-provoking   romance  \
And Then There Were None                              0.542501  0.718039   
The Lion, the Witch and the Wardrobe                  0.509143  0.722981   
Harry Potter and the Chamber of Secrets               0.562555  0.772843   
The Catcher in the Rye                                0.631479  0.753063   
Lolita                                                0.593820  0.739397   
...                                                        ...       ...   
The Human Mind                                        0.561223  0.576303   
This Little Chick                                     0.535775  0.747039   
View with a Grain of Sand: Selected Poems             0.606666  0.729281   
Rogue Male                                            0.584509  0.700178   
Stokes Field Guide to Birds: Western Region           0.520172  0.673966   

                                                 cozy       fun  
And Then There Were None                     0.184678  0.620200  
The Lion, the Witch and the Wardrobe         0.175687  0.608017  
Harry Potter and the Chamber of Secrets      0.243108  0.701532  
The Catcher in the Rye                       0.204740  0.633474  
Lolita                                       0.182729  0.571301  
...                                               ...       ...  
The Human Mind                               0.181041  0.539014  
This Little Chick                            0.275895  0.692513  
View with a Grain of Sand: Selected Poems    0.163858  0.629822  
Rogue Male                                   0.235087  0.614593  
Stokes Field Guide to Birds: Western Region  0.241383  0.682278  

[1365 rows x 9 columns]

## option 2 -> process stopped, we will stick with the 1365 words

In [217]:
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity


#df = pd.read_csv("C:/Users/chiar/Downloads/final500 (2).csv")
df = pd.read_csv("C:/Users/chiar/Downloads/final10000.csv")
df =  df.dropna(subset=['review/text'])            
df = df.dropna(subset=['num_pages'])
df = df.drop(columns=['nan'])
df = df.drop(columns=['Unnamed: 0'])
#df = df.drop(df[df["review/text"] == "['nan']"].index)
#df = df.drop(columns=['Description'])
#df = df.drop(columns=['review/text'])
#df.to_csv("final_1365books.csv")

df

Title  \
0                                  A Tale of Two Cities   
3                              And Then There Were None   
6                  The Lion, the Witch and the Wardrobe   
9                                     The Da Vinci Code   
10              Harry Potter and the Chamber of Secrets   
...                                                 ...   
9494                                     Yurara, Vol. 1   
9495                                     Yurara, Vol. 2   
9496                                               Finn   
9497  God, Justice, and Society: Aspects of Law and ...   
9498  No Holds Barred Fighting: The Clinch: Offensiv...   

                         Author  \
0               Charles Dickens   
3               Agatha Christie   
6                   C. S. Lewis   
9                     Dan Brown   
10                J. K. Rowling   
...                         ...   
9494               Chika Shiomi   
9495               Chika Shiomi   
9496                 Jon Clinch   
9497          Jonathan Burnside   
9498  Mark Hatmaker,Doug Werner   

                                                  Genre  \
0                                    Historical fiction   
3                                               Mystery   
6                           Fantasy, Children's fiction   
9                                      Mystery thriller   
10                                              Fantasy   
...                                                 ...   
9494  Sequential Art,Manga,Fantasy,Sequential Art,Gr...   
9495  Sequential Art,Manga,Fantasy,Romance,Sequentia...   
9496  Fiction,Historical,Historical Fiction,Historic...   
9497                              Law,Religion,Theology   
9498                                Combat,Martial Arts   

                                            Description  num_pages  \
0     Edited with an Introduction and Notes by Richa...      489.0   
3                                                   NaN      264.0   
6                                                   NaN      206.0   
9     While in Paris on business, Harvard symbologis...      489.0   
10    When the Chamber of Secrets is opened again at...      352.0   
...                                                 ...        ...   
9494                                                NaN      200.0   
9495                                                NaN      192.0   
9496                                                NaN      287.0   
9497                                                NaN      542.0   
9498                                                NaN      224.0   

                                            review/text  
0                                               ['nan']  
3     ['book', 'one', 'ten', 'read', 'mystery', 'kil...  
6     ['lewis', 'world', 'witch', 'fantasy', 'wardro...  
9                                               ['nan']  
10    ['book', 'potter', 'harry', 'fiction', 'scienc...  
...                                                 ...  
9494                                            ['nan']  
9495                                            ['nan']  
9496                                            ['nan']  
9497                                            ['nan']  
9498                                            ['nan']  

[9348 rows x 6 columns]

In [ ]:
img = pd.read_csv("C:/Users/chiar/Downloads/moods (1).csv")
img["image"] = [0, 1, 2, 3, 4, 5, 6, 7, 8]

#getting the summary vector for the books' descriptions
for index, row in df.iterrows():
    review_text = row['review/text']
    if review_text != str(['nan']):
        review_text = review_text.strip("[] ' ' \"  ")
        review_text = review_text.replace("'", "")
        review_text = review_text.replace(" ", "")
        li = list(review_text.split(","))
        li = [word for word in li if word in model]
        summary_vector = sum(model[word] for word in li) / len(li)
        df.at[index, 'review/text'] = summary_vector
df 

#getting the summary vector for the images descriptions
for index, row in img.iterrows():
    review_text = row['vector']
    if review_text != str(['nan']):
        li = list(review_text.split(","))
        li = [word for word in li if word in model]
        summary_vector = sum(model[word] for word in li) / len(li)
        img.at[index, 'vector'] = summary_vector

img

def create_similarity_matrix(book_description_vectors, mood_vectors):
    similarity_matrix = np.zeros((len(book_description_vectors), len(mood_vectors)))
    for i, book_vector in enumerate(book_description_vectors):
        for j, mood_vector in enumerate(mood_vectors):
            if isinstance(book_vector, str) and book_vector == "['nan']":
                similarity_matrix[i, j] = 0
            else:
                #this check now becomes useless because we removed those earlier
                if np.isnan(book_vector).any() or np.isnan(mood_vector).any():
                    similarity_matrix[i, j] = 0
                else:
                    similarity_score = cosine_similarity([book_vector], [mood_vector])[0][0]
                    similarity_matrix[i, j] = similarity_score
    return similarity_matrix

mood_vectors = img['vector']
book_vectors = df['review/text']
book_titles = df['Title'].tolist()
similarity_matrix = create_similarity_matrix(book_vectors, mood_vectors)

matrix = pd.DataFrame(similarity_matrix, index=book_titles, columns=["happy", "sad", "misterious", "spring", "adventurous", "thought-provoking", "romance", "cozy", "fun"])
matrix

non_zero_rows = matrix[(matrix > 0.0000000).any(axis=1)]
#non_zero_rows.to_csv("mood_matrix_reviews.csv")

non_zero_rows